File: 3147-taking-maximum-energy-from-the-mystic-dungeon/3147-taking-maximum-energy-from-the-mystic-dungeon.cpp
class Solution {
public:
    int maximumEnergy(vector<int>& energy, int k) {
        int n = energy.size();
        vector<int> dp = energy;
        for (int i = n - k - 1; i >= 0; --i) {
            dp[i] += dp[i + k];
        }
        return *max_element(dp.begin(), dp.end());
    }
};

File: 3147-taking-maximum-energy-from-the-mystic-dungeon/3147-taking-maximum-energy-from-the-mystic-dungeon.py
class Solution:
    def maximumEnergy(self, energy: List[int], k: int) -> int:
        n = len(energy)
        dp = energy[:]

        for i in range(n-k-1, -1, -1):
            dp[i] += dp[i+k]

        return max(dp)